processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metricstransformprocessor
+
+import (
+	"context"
+	"log"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+type metricsTransformProcessor struct {
+	cfg        *Config
+	next       consumer.MetricsConsumer
+	metricname string
+	action     ConfigAction
+	newname    string
+	operations []Operation
+}
+
+var _ component.MetricsProcessor = (*metricsTransformProcessor)(nil)
+
+func newMetricsTransformProcessor(next consumer.MetricsConsumer, cfg *Config) (*metricsTransformProcessor, error) {
+	return &metricsTransformProcessor{
+		cfg:        cfg,
+		next:       next,
+		metricname: cfg.MetricName,
+		action:     cfg.Action,
+		newname:    cfg.NewName,
+		operations: cfg.Operations,
+	}, nil
+}
+
+// GetCapabilities returns the Capabilities assocciated with the resource processor.
+func (mtp *metricsTransformProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: false}
```

Comment:
I think this is where I am a bit confused. When the ConsumeMetrics function calls the transform function, the argument is passed in as value instead of reference. Is this equivalent to making a copy, operating on a copy, and then passing back the copy? In this case, I think the original data is not modified? I am probably understanding it wrong, but more clarification would definitely help.

Reply:
The argument in this case is of type `pdata.MetricData` which is an interface type. Interfaces in Go, like slices & maps (but not arrays), are essentially always passed as pointers.

Here's a good article that explains this better: https://medium.com/@saiyerram/go-interfaces-pointers-4d1d98d5c9c6

Also note if you were copying that whole dataset that would be potentially very expensive (even if it gets allocated on the stack), so we definitely would want to avoid doing that.

Reply:
Got it I see!